### ISSUE TITLE
InvalidImageName when deploying applications

### ISSUE THREAD

**Issue by kwoodson** [openshift/openshift-azure#159]
When deploying a cluster I need to test whether the registry is working.  I deployed from /console the cakephp-mysql-persistent application.  It shows up as "InvalidImageName":
```
[root@master-000000 ~]# oc get pods
NAME                               READY     STATUS                  RESTARTS   AGE
cakephp-mysql-persistent-2-build   0/1       Init:InvalidImageName   0          57s
mysql-1-deploy                     0/1       InvalidImageName        0          5m
```

I checked the build pod and it has a broken image format:
```
[root@master-000000 ~]# oc get pods cakephp-mysql-persistent-2-build -o yaml | grep image
    image: 'docker.io/openshift/origin-:'
    imagePullPolicy: IfNotPresent
  imagePullSecrets:
    image: 'docker.io/openshift/origin-:'
    imagePullPolicy: IfNotPresent
    image: 'docker.io/openshift/origin-:'
    imagePullPolicy: IfNotPresent
  - image: 'docker.io/openshift/origin-:'
    imageID: ""
  - image: 'docker.io/openshift/origin-:'
    imageID: ""
        message: 'Failed to apply default image tag "docker.io/openshift/origin-:":
          couldn''t parse image reference "docker.io/openshift/origin-:": invalid
  - image: 'docker.io/openshift/origin-:'
    imageID: ""
```

I checked the master-config.yaml and the image string looks like this:
```
format: "docker.io/openshift/origin-:"
```

This leads me to believe that the string is not being handled correctly when it gets laid down.  This led me to this code:
https://github.com/openshift/openshift-azure/blob/master/pkg/arm/data/master-startup.sh#L171
https://github.com/openshift/openshift-azure/blob/master/pkg/config/generate.go#L46

I'll look for a few more minutes and will update.

@kargakis @jim-minter @mjudeikis, Any ideas on this one?

**Comment by kwoodson** [comment]
Fixing the value manually in each master's master-config.yaml and restarting the controller pods did not fix the issue.  Still seeing build issues.

**Comment by kwoodson** [comment]
@pweil-  suggested that the shell script is interpolating the `${component}` and the `${version}`.

**Comment by kwoodson** [comment]
We cannot just `'EOF'` as the `$(hostname)` line is also being interpolated correctly.

https://github.com/openshift/openshift-azure/blob/master/pkg/arm/data/master-startup.sh#L164
and
https://github.com/openshift/openshift-azure/blob/master/pkg/arm/data/master-startup.sh#L171

**Comment by kwoodson** [comment]
Main issue is that the `${component}` and `${version}` are being interpolated in bash before writing to `master-config.yaml`.  This breaks image pulling as the string appears like this:
```
#bad
'docker.io/openshift/origin-:'
#good
'docker.io/openshift/origin-${component}:${version}'
```
Since there are `$(hostname)` calls as well we cannot just `'EOF'` the command that writes the config file.  This is blocking testing of my MSI registry changes as I cannot build and push to the registry.

A few options here: @kargakis @pweil- @jim-minter @mjudeikis 
- change `| quote` to `| escape`.  Add an escape filter to `\$` in the variable when writing it.
- change `$(hostname)` to the hostname of the box and update `'EOF'`.
- write a `CHANGEME` to master-config.yaml and then sed it after the fact with the correct image format
- other ideas?

Thanks!

**Comment by kargakis** [comment]
I would try escaping first

**Comment by pweil-** [comment]
> I would try escaping first

:+1:

**Comment by kwoodson** [comment]
Escaping worked great.  Fix included in #144

**Comment by kwoodson** [comment]
```
grep format /etc/origin/master/master-config.yaml 
  format: docker.io/openshift/origin-${component}:${version}
```

**Comment by kwoodson** [comment]
Fixed in #144 
/close
